sky/clouds/service_catalog/data_fetchers/fetch_gcp.py

```diff
@@ -1,579 +1,308 @@
+    # Get the SKUs from the GCP API.
+    cb = discovery.build('cloudbilling', 'v1')
+    service_name = 'services/' + service_id
+
+    skus = []
+    page_token = ''
+    while True:
+        if page_token == '':
+            response = cb.services().skus().list(parent=service_name).execute()
         else:
-            return float(memory_str)
-
-    pattern = re.compile(r'\$?(.*?)\s?/')
-
-    def parse_price(price_str: str) -> float:
-        if NOT_AVAILABLE_STR in price_str:
-            return float('nan')
-        try:
-            price = float(price_str[1:])
-        except ValueError:
-            price_match = re.search(pattern, price_str)
-            if price_match is None:
-                raise ValueError(f'Cannot parse price: {price_str}') from None
-            price = float(price_match.group(1))
-        return price
-
-    # Parse the prices.
-    df['Price'] = df['Price'].apply(parse_price)
-    df['SpotPrice'] = df['SpotPrice'].apply(parse_price)
-
-    # Remove unsupported regions.
-    df = df[~df['Price'].isna()]
-    df = df[~df['SpotPrice'].isna()]
-
-    # E.g., m2-ultramem instances will be skipped because their spot prices
-    # are not available.
-    if df.empty:
-        return None
+            response = cb.services().skus().list(
+                parent=service_name, pageToken=page_token).execute()
+        skus += response['skus']
+        page_token = response['nextPageToken']
+        if not page_token:
+            break
+
+    # Prune unnecessary SKUs.
+    new_skus = []
+    for sku in skus:
+        # Prune SKUs that are not Compute (i.e., Storage, Network, and License).
+        if sku['category']['resourceFamily'] != 'Compute':
+            continue
+        # Prune PD snapshot egress and VM state.
+        if sku['category']['resourceGroup'] in ['PdSnapshotEgress', 'VmState']:
+            continue
+        # Prune commitment SKUs.
+        if sku['category']['usageType'] not in ['OnDemand', 'Preemptible']:
+            continue
+        # Prune custom SKUs.
+        if 'custom' in sku['description'].lower():
+            continue
+        # Prune premium SKUs.
+        if 'premium' in sku['description'].lower():
+            continue
+        # Prune reserved SKUs.
+        if 'reserved' in sku['description'].lower():
+            continue
+        # Prune sole-tenant SKUs.
+        # See https://cloud.google.com/compute/docs/nodes/sole-tenant-nodes
+        if 'sole tenancy' in sku['description'].lower():
+            continue
+        new_skus.append(sku)
+    return new_skus
+
+
+def _get_unit_price(sku: Dict[str, Any]) -> float:
+    pricing_info = sku['pricingInfo'][0]['pricingExpression']
+    unit_price = pricing_info['tieredRates'][0]['unitPrice']
+    units = int(unit_price['units'])
+    nanos = unit_price['nanos'] / 1e9
+    return units + nanos
+
+
+def get_vm_df(skus: List[Dict[str, Any]]) -> pd.DataFrame:
+    df = pd.read_csv(GCP_METADATA_URL + 'instances.csv')
```

Comment:
Got it. I'm just a little bit worried that the metadata part may change faster than we thought. I did some quick diff on the old and new `vms.csv`. It seems to me the change is actually not small. ~90 out of 153 instance types
' supported zones are changed. See details in the gist https://gist.github.com/infwinston/c62c488f9c61e8f34780053d4be16fc3
or you think generating them on the fly can cause too much trouble?

Reply:
@infwinston Yeah it seems there have been many updates in GCE recently, including the introduction of the new [C3 instance types](https://cloud.google.com/blog/products/compute/introducing-c3-machines-with-googles-custom-intel-ipu).

I think there's a tradeoff between freshness and robustness of the catalog. If we create the metadata on the fly, then the generated catalog may not be robust because
1) The GCP pricing API may not work for new instance types,
2) The GCP pricing API may not work for new regions/zones,
3) The [repository](https://github.com/Cyclenerd/google-cloud-pricing-cost-calculator) we fetch the data from may include errors on new VMs/regions/zones,
4) Our script will break if the repository changes its directory structure or file names.
Besides, we manually maintain the TPU metadata and do not have a good way to automate it.

In addition, again, I think some staleness is acceptable as we don't have users interested in using the new instance types/regions/zones.

Overall, I believe the manually updating the metadata + automatically updating the prices is a better solution at the moment.

Reply:
@infwinston WDYT?

Reply:
Sorry @WoosukKwon Got it. Correct me if I'm wrong. I guess that those concerns may be fixed by pinning the version of the repository. so avoiding storing big CSV files in our repo is still possible? And we may manually bump the version if any user request the latest catalog.
Does this sound like an option to you? let me know if you still think it's too complicated. My concern is it seems unnecessary to store big CSVs that don't require version control in our git repo.